### ISSUE TITLE
Compile error with Elixir 1.3.0-rc.0

### ISSUE THREAD

**Issue by chrismcg** [boldpoker/edeliver#94]
I tried upgrading my build box to 1.3.0-rc.0 (Erlang is 18.3) and ran into the following error:

Note, I had to edit my local deps to add `--force` here https://github.com/boldpoker/edeliver/blob/master/libexec/erlang#L123 to get this far.

`mix edeliver build release --branch=staging --auto-version=git-revision` gives me:

```
BUILDING RELEASE OF OSTALI APP ON BUILD HOST

-----> Authorizing hosts
-----> Ensuring hosts are ready to accept git pushes
-----> Pushing new commits with git to: vagrant@elixir_build_box
-----> Resetting remote hosts to 2d9b517db0143ebca280c07df1ef326ca42fb110
-----> Cleaning generated files from last build
-----> Copying 'config/ostali.conf' file to build host
-----> Fetching / Updating dependencies
-----> Compiling sources
** (Mix) Cannot execute task because the project was not yet compiled. When build_embedded is set to true, "MIX_ENV=prod mix compile" must be explicitly executed

A remote command failed on:

  vagrant@elixir_build_box

Output of the command is shown above and the command executed
on that host is printed below for debugging purposes:

FAILED with texit status 1:

    [ -f ~/.profile ] && source ~/.profile
    set -e
    cd /tmp/erlang/ostali/builds
    if [ "mix" = "rebar" ]; then
      echo "using rebar to compile files"
      [[ "" != "true" ]] && ./rebar  clean skip_deps=true || :
      ./rebar  compile
    elif [ "mix" = "mix" ] && [ "mix" = "mix" ]; then
      if [[ "" = "true" ]]; then
        APP="ostali" MIX_ENV="prod" AUTO_VERSION="" SKIP_RELUP_MODIFICATIONS="" RELUP_MODIFICATION_MODULE="" mix do release.version git-revision , compile
      else
        APP="ostali" MIX_ENV="prod" AUTO_VERSION="" SKIP_RELUP_MODIFICATIONS="" RELUP_MODIFICATION_MODULE="" mix do clean, release.version git-revision , compile
      fi
    elif [ "mix" = "mix" ]; then
      echo "using mix to compile files"
      if [[ "" = "true" ]]; then
        APP="ostali" MIX_ENV="prod" AUTO_VERSION="" SKIP_RELUP_MODIFICATIONS="" RELUP_MODIFICATION_MODULE="" mix do deps.compile, release.version git-revision , compile
      else
        APP="ostali" MIX_ENV="prod" AUTO_VERSION="" SKIP_RELUP_MODIFICATIONS="" RELUP_MODIFICATION_MODULE="" mix do clean, deps.compile, release.version git-revision , compile
      fi
    fi
```

If I use:

`mix edeliver build release --branch=staging --skip-mix-clean`

Then it builds the release just fine. Don't have to much time to dig into this right now so going back to 1.2.5. If I do find out how to get round it I'll update.

**Comment by chrismcg** [comment]
Commenting out https://github.com/boldpoker/edeliver/blob/master/libexec/erlang#L168-L173 so the `mix do` list included a `deps.compile` worked round the problem for me.

**Comment by bharendt** [comment]
Thanks for the detailed information. At a first sight, it looks like the `mix compile` task does not compile the dependencies automatically any more for that elixir/mix version. I think we could add the `deps.compile` task also to the other branches of that code, because it should not have any negative effects. It should actually not recompile them (at the `mix compile` step) if they are then already compiled. But would be nice if we could verify in the mix code that this behaviour changed, before updading the edeliver build commands.
I think it would also have no drawbacks to add the `--force` flag to the `local.rebar` command. But would be nice to know why it failed. Do you maybe still have the output for the failed `mix local.rebar` command withtout the `--force` flag? 
Btw: using `--skip-mix-clean` would actually prevent `--auto-version` to work correctly. It is required to force the re-generation of the `your_app.app` file with the updated version.

**Comment by chrismcg** [comment]
I needed to add --force because hex was asking for [Yn] answer which I discovered when I ran with `--verbose`

```
-----> Fetching / Updating dependencies
using mix to fetch and update deps
/home/vagrant/.asdf/installs/elixir/1.3.0-rc.0/.mix/rebar already exists, overwrite? [Yn]
* creating /home/vagrant/.asdf/installs/elixir/1.3.0-rc.0/.mix/rebar
/home/vagrant/.asdf/installs/elixir/1.3.0-rc.0/.mix/rebar3 already exists, overwrite? [Yn]
```

It did indeed complain when I first tried `--auto-version` and `--skip-mix-clean`, so I did it without the `--auto-version` as an experiment. I just mentioned it in case it was useful for you in debugging :)

**Comment by bharendt** [comment]
Ok, thanks. I will then add the `--force` flag...

**Comment by chrismcg** [comment]
I'm not sure what's changed in Elixir to cause the issue, there's nothing obvious in the Changelog. Perhaps worth asking on IRC or the core mailing list.

(btw adding the deps.compile to the other branch will make both branches the same afaics.)

**Comment by bharendt** [comment]
Thanks for the hint! Changed that and removed that other branch completely. Will try to verify that change at the elixir/mix code tomorrow, and of course any help with that is appreciated.

**Comment by bharendt** [comment]
I was not able to reproduce that the dependencies are not automatically compiled any more when running `mix compile` and using elixir version `1.3.0-rc.0`. Because `mix deps.compile` is time consuming (because it checks all dependencies even if they don't need to be compiled) I will revert [that commit](https://github.com/boldpoker/edeliver/commit/3056131069c2462535386c16fe1ccfb4dceb90c1) for the upcomming edeliver version `1.2.9` and until new information indicate that an explicit compilation of the dependencies is still required.

**Comment by aaronjensen** [comment]
Just got bit by this. deps are not automatically compiled in elixir 1.3 afaict

**Comment by aaronjensen** [comment]
Related: https://github.com/phoenixframework/phoenix/issues/1766

**Comment by aaronjensen** [comment]
So this was an elixir bug and will be fixed in 1.3.1. I've worked around it w/ a fork for now. Not sure how you'd want to proceed @bharendt, but edeliver won't work w/ elixir 1.3 at the moment out of the box.

**Comment by bharendt** [comment]
Compiling the dependencies explicitly for mix version `1.3.0` is included in [edeliver version 1.2.10](https://hex.pm/packages/edeliver/1.2.10)

**Comment by beno** [comment]
I am running into this issue trying to deploy a mix umbrella app, with the same output as above. It is easy to see why the last branch of the if statement is never reached.

Running elixir/mix 1.3.1

**Comment by st23am** [comment]
I'm also running into this with an umbrella app..

**Comment by netflakes** [comment]
I have the same issue with Elixir 1.3.2

**Comment by brienw** [comment]
It looks like the fix didn't make it into elixir/mix 1.3.3 either: https://github.com/elixir-lang/elixir/issues/5094#issuecomment-248408478

**Comment by brienw** [comment]
I opened https://github.com/boldpoker/edeliver/pull/165 to work around this again for the later versions.
